crates/extensions/tedge_mqtt_ext/src/lib.rs

```diff
@@ -134,12 +134,18 @@ impl FromPeers {
         &mut self,
         outgoing_mqtt: &mut mpsc::UnboundedSender<MqttMessage>,
     ) -> Result<(), RuntimeError> {
-        loop {
-            match self.try_recv().await {
-                Ok(Some(message)) => outgoing_mqtt.send(message).await.map_err(Box::new)?,
-                Ok(None) | Err(RuntimeRequest::Shutdown) => break Ok(()),
-            }
+        while let Ok(Some(message)) = self.try_recv().await {
+            outgoing_mqtt.send(message).await.map_err(Box::new)?;
         }
+
+        // On shutdown, first close input so no new messages can be pushed
+        self.input_receiver.close_input();
```

Comment:
Shouldn't the `incoming_mqtt` channel should also be closed to avoid reading any more messages from the broker as well? For that, these `relay_xxx` methods would need to take the mutable `Connection` itself, instead of the `published` or `received` channels from it.

Reply:
Unfortunately, it would be pointless to close the `incoming_mqtt` channel as the underlying TCP connection to the broker is bi-directional: as we want to publish a latest batch of messages; we have to keep the connection open.